File: PrefixSum/3652. Best Time to Buy and Sell Stock using Strategy/solution.py
class Solution:
    def maxProfit(self, prices: List[int], strategy: List[int], k: int) -> int:
        n = len(prices)

        prefix_prices = list(accumulate(prices, initial=0))
        presum = [0]
        for p, st in zip(prices, strategy):
            presum.append(presum[-1] + p*st)

        j = 0
        res = presum[n]
        for i in range(n-k+1):
            while j < n and j - i < k:
                j += 1

            # Set the first k / 2 elements to 0 (hold).
            # Set the last k / 2 elements to 1 (sell).
            # only earned second-half
            mid = prefix_prices[j]-prefix_prices[j-k//2]
            res = max(res, presum[i] + mid + presum[n]-presum[j])
        return res

File: PrefixSum/3652. Best Time to Buy and Sell Stock using Strategy/solution.cpp
#include <vector>
#include <numeric>

using namespace std;

class Solution {
public:
    long long maxProfit(vector<int>& prices, vector<int>& strategy, int k) {
        int n = prices.size();
        vector<long long> presum(n+1, 0);
        vector<long long> preprof(n+1, 0);
        long long res = 0;
        for (int i=0; i<n; i++) {
            res += prices[i] * strategy[i];
            presum[i+1] = presum[i] + prices[i];
            preprof[i+1] = preprof[i] + prices[i] * strategy[i];
        }

        for (int i=k-1; i<n; i++) {
            long long operation = presum[i+1] - presum[i+1-k/2];
            long long right = preprof[n] - preprof[i+1];
            long long left = preprof[i+1-k];
            res = max(res, left + operation + right);
        }
        return res;
    }
};
